### PR TITLE
bug fix #136: ckpt saving logic changed

### LINKED ISSUE #136: expected behavior for `keep_last_ckpts = -1`
Hi @juliakreutzer,

I wanted to save ckpts at every validation step regardless of early-stopping-metric score, so I set `keep_last_ckpts = -1`, according to the description here:

https://github.com/joeynmt/joeynmt/blob/46b2fe3b05638728413ee5bae6a347411175c3c5/configs/small.yaml#L67

But joeynmt didn't save ckpts at all in that case. Actually, the TrainManager doesn't call `_save_checkpoint()` func if `keep_last_ckpts` is less than or equal to zero (queue with infinite length: https://docs.python.org/3/library/queue.html).
https://github.com/joeynmt/joeynmt/blob/46b2fe3b05638728413ee5bae6a347411175c3c5/joeynmt/training.py#L98-L99
https://github.com/joeynmt/joeynmt/blob/46b2fe3b05638728413ee5bae6a347411175c3c5/joeynmt/training.py#L544-L547

What is the expected behavior? You indeed intended no save action if `keep_last_ckpts = -1`, that is, the description in config was wrong or can we change the code so that ckpts will be saved every time if `keep_last_ckpts = -1`?

### PR DESCRIPTION
**Pull request by may-**
- bug in ckpt saving/deleting logic fixed #136
  - config option name changed: `keep_last_ckpts` -> `keep_best_ckpts`
  - `save_latest_ckpt` flag is removed; now this is a default behaviour, the `latest.ckpt` symlink is **always** created.
  - now `keep_best_ckpts=-1` works :)
- new stopping criterion `updates` (in addition to `epochs`)
  - `epochs` -> will be reset in resumed training
  - `updates` -> the max. number of updates, which will be inherited from the previous training process even in resumed session
- default vocab file paths: `<model_dir>/src_vocab.txt` and `<model_dir>/trg_vocab.txt` also in test mode (not only in translation mode), in case vocab files are not explicitly specified in config.
- `torchtext.legacy.data.BucketIterator` is replaced with `torchtext.legacy.data.Iterator`.
  - preparation in order to get rid of torchtext in near future

### CODE REVIEW

File: joeynmt/helpers.py [hunk 1]
```diff
@@ -119,8 +118,8 @@ def subsequent_mask(size: int) -> Tensor:
     :param size: size of mask (2nd and 3rd dim)
     :return: Tensor with 0s and 1s of shape (1, size, size)
     """
-    mask = np.triu(np.ones((1, size, size)), k=1).astype('uint8')
-    return torch.from_numpy(mask) == 0
+    ones = torch.ones(size, size, dtype=torch.bool)
+    return torch.tril(ones, out=ones).unsqueeze(0)
```
Comment (juliakreutzer) [inline joeynmt/helpers.py:122]: nice!

File: joeynmt/helpers.py [hunk 2]
```diff
@@ -323,38 +322,45 @@ def freeze_params(module: nn.Module) -> None:
...
-        if e.errno == errno.EEXIST:
-            os.remove(link_name)
-            os.symlink(target, link_name)
-        else:
-            raise e
+        logger.info('delete %s', to_delete.as_posix())
+        to_delete.unlink()
+
+    except FileNotFoundError as e:
+        #logger = logging.getLogger(__name__)
```
Comment (juliakreutzer) [inline joeynmt/helpers.py:337]: this comment can be removed, right?

File: requirements.txt [hunk 3]
```diff
@@ -1,15 +1,15 @@
 future
 pillow 
-numpy==1.20.1
```
Comment (juliakreutzer) [inline requirements.txt:3]: did you test it with newer versions?

**Review by juliakreutzer — APPROVED** [review]
Looks great, thank you so much! :100:

**Comment by may-** [comment]
Hi @juliakreutzer

**Comment by may-** [comment]
Hi @juliakreutzer ,

results on iwslt14 de-en (sentencepiece joint bpe8000)

||dev|test|
|:-|:-:|:-:|
|BLEU|35.30|35.18|

I tested this pull request on the following environment:

- python 3.9
- cuda 10.2
- pytorch 1.9.0
- torchtext 0.10.0
- numpy 1.20.1
- pylint 2.6.0

I also cherry-picked #149 . If you see any problem, please let me know!

**Comment by juliakreutzer** [comment]
thank you!!
